After upgrading from 1.1.2 to 1.1.3

**Describe the bug**
After upgrading the library it started to fail with:

```
Done parsing Play: Run cluster deployment on prepared hypervisors (1) **********
Traceback (most recent call last):
  File "/usr/local/bin/ansible-playbook-grapher", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/dist-packages/ansibleplaybookgrapher/cli.py", line 249, in main
    cli.run()
  File "/usr/local/lib/python3.8/dist-packages/ansibleplaybookgrapher/cli.py", line 71, in run
    svg_path = renderer.render(
  File "/usr/local/lib/python3.8/dist-packages/ansibleplaybookgrapher/renderer.py", line 114, in render
    self._convert_to_graphviz()
  File "/usr/local/lib/python3.8/dist-packages/ansibleplaybookgrapher/renderer.py", line 391, in _convert_to_graphviz
    self.render_node(
  File "/usr/local/lib/python3.8/dist-packages/ansibleplaybookgrapher/renderer.py", line 153, in render_node
    self.render_block(
  File "/usr/local/lib/python3.8/dist-packages/ansibleplaybookgrapher/renderer.py", line 244, in render_block
    self.render_node(
  File "/usr/local/lib/python3.8/dist-packages/ansibleplaybookgrapher/renderer.py", line 161, in render_node
    self.render_role(
  File "/usr/local/lib/python3.8/dist-packages/ansibleplaybookgrapher/renderer.py", line 295, in render_role
    role_color = ":".join([c[0] for c in colors])
  File "/usr/local/lib/python3.8/dist-packages/ansibleplaybookgrapher/renderer.py", line 295, in <listcomp>
    role_color = ":".join([c[0] for c in colors])
TypeError: 'NoneType' object is not subscriptable
Error: Process completed with exit code 1.
```

**Versions:**
Outputs of :
 - `ansible --version`
 ansible-5.1.0 ansible-compat-2.2.0 ansible-core-2.12.8
 - `ansible-playbook-grapher --version`
 Error from 1.1.2(works) to 1.1.3(broken)

--- a/ansibleplaybookgrapher/graph.py
+++ b/ansibleplaybookgrapher/graph.py
@@ -24,7 +24,14 @@ class Node:
     A node in the graph. Everything of the final graph is a node: playbook, plays, edges, tasks and roles.
     """
 
-    def __init__(self, node_name: str, node_id: str, when: str = "", raw_object=None, parent: 'Node' = None):
+    def __init__(
+        self,
+        node_name: str,
+        node_id: str,
+        when: str = "",
+        raw_object=None,
+        parent: "Node" = None,
+    ):
         """
 
         :param node_name: The name of the node
@@ -51,7 +58,7 @@ class Node:
         if self.raw_object and self.raw_object.get_ds():
             self.path, self.line, self.column = self.raw_object.get_ds().ansible_pos
 
-    def get_first_parent_matching_type(self, node_type: Type) -> 'Type':
+    def get_first_parent_matching_type(self, node_type: Type) -> "Type":
         """
         Get the first parent of this node matching the given type
         :return:
@@ -66,7 +73,7 @@ class Node:
         raise ValueError(f"No parent of type {node_type} found for {self}")
 
     def __repr__(self):
-        return f"{type(self).__name__}(name='{self.name}',id='{self.id}')"
+        return f"{type(self).__name__}(name='{self.name}', id='{self.id}')"
 
     def __eq__(self, other):
         return self.id == other.id
@@ -89,7 +96,7 @@ class CompositeNode(Node):
         node_id: str,
         when: str = "",
         raw_object=None,
-        parent: 'Node' = None,
+        parent: "Node" = None,
         supported_compositions: List[str] = None,
     ):
         """
@@ -176,8 +183,17 @@ class CompositeTasksNode(CompositeNode):
     A special composite node which only support adding "tasks". Useful for block and role
     """
 
-    def __init__(self, node_name: str, node_id: str, when: str = "", raw_object=None, parent: 'Node' = None):
-        super().__init__(node_name, node_id, when=when, raw_object=raw_object, parent=parent)
+    def __init__(
+        self,
+        node_name: str,
+        node_id: str,
+        when: str = "",
+        raw_object=None,
+        parent: "Node" = None,
+    ):
+        super().__init__(
+            node_name, node_id, when=when, raw_object=raw_object, parent=parent
+        )
         self._supported_compositions = ["tasks"]
 
     def add_node(self, target_composition: str, node: Node):
@@ -232,31 +248,25 @@ class PlaybookNode(CompositeNode):
         """
         return self._compositions["plays"]
 
-    def roles_usage(self) -> Dict["RoleNode", Set[Node]]:
+    def roles_usage(self) -> Dict["RoleNode", List[Node]]:
         """
         For each role in the graph, return the plays that reference the role
         FIXME: Review this implementation. It may not be the most efficient way, but it's ok for the moment
         :return: A dict with key as role node and value the list of plays
         """
 
-        usages = defaultdict(set)
+        usages = defaultdict(list)
         links = self.links_structure()
 
         for node, linked_nodes in links.items():
             for linked_node in linked_nodes:
                 if isinstance(linked_node, RoleNode):
                     if isinstance(node, PlayNode):
-                        usages[linked_node].add(node)
+                        usages[linked_node].append(node)
                     else:
-                        usages[linked_node].add(node.get_first_parent_matching_type(PlayNode))
-
-        # In case a role is used by another role, replace it by the play associated with using role (transitivity)
-        for usages_set in usages.values():
-            for node in usages_set.copy():
-                for r in usages:
-                    if node.id == r.id:
-                        usages_set.remove(node)
-                        usages_set.update(usages[r])
+                        usages[linked_node].append(
+                            node.get_first_parent_matching_type(PlayNode)
+                        )
 
         return usages
 
@@ -276,7 +286,7 @@ class PlayNode(CompositeNode):
         node_id: str = None,
         when: str = "",
         raw_object=None,
-        parent: 'Node' = None,
+        parent: "Node" = None,
         hosts: List[str] = None,
     ):
         """
@@ -317,7 +327,12 @@ class BlockNode(CompositeTasksNode):
     """
 
     def __init__(
-        self, node_name: str, node_id: str = None, when: str = "", raw_object=None, parent: 'Node' = None
+        self,
+        node_name: str,
+        node_id: str = None,
+        when: str = "",
+        raw_object=None,
+        parent: "Node" = None,
     ):
         super().__init__(
             node_name,
@@ -334,7 +349,12 @@ class TaskNode(Node):
     """
 
     def __init__(
-        self, node_name: str, node_id: str = None, when: str = "", raw_object=None, parent: 'Node' = None
+        self,
+        node_name: str,
+        node_id: str = None,
+        when: str = "",
+        raw_object=None,
+        parent: "Node" = None,
     ):
         """
 
@@ -343,7 +363,11 @@ class TaskNode(Node):
         :param raw_object:
         """
         super().__init__(
-            node_name, node_id or generate_id("task_"), when=when, raw_object=raw_object, parent=parent
+            node_name,
+            node_id or generate_id("task_"),
+            when=when,
+            raw_object=raw_object,
+            parent=parent,
         )
 
 
@@ -358,7 +382,7 @@ class RoleNode(CompositeTasksNode):
         node_id: str = None,
         when: str = "",
         raw_object=None,
-        parent: 'Node' = None,
+        parent: "Node" = None,
         include_role: bool = False,
     ):
         """
@@ -369,7 +393,11 @@ class RoleNode(CompositeTasksNode):
         """
         self.include_role = include_role
         super().__init__(
-            node_name, node_id or generate_id("role_"), when=when, raw_object=raw_object, parent=parent
+            node_name,
+            node_id or generate_id("role_"),
+            when=when,
+            raw_object=raw_object,
+            parent=parent,
         )
 
     def retrieve_position(self):

--- a/ansibleplaybookgrapher/graph.py
+++ b/ansibleplaybookgrapher/graph.py
@@ -14,8 +14,7 @@
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 import os
 from collections import defaultdict
-from typing import Dict, List, ItemsView, Set
-from collections import defaultdict
+from typing import Dict, List, ItemsView, Set, Type
 
 from ansibleplaybookgrapher.utils import generate_id
 
@@ -25,7 +24,7 @@ class Node:
     A node in the graph. Everything of the final graph is a node: playbook, plays, edges, tasks and roles.
     """
 
-    def __init__(self, node_name: str, node_id: str, when: str = "", raw_object=None):
+    def __init__(self, node_name: str, node_id: str, when: str = "", raw_object=None, parent: 'Node' = None):
         """
 
         :param node_name: The name of the node
@@ -33,8 +32,10 @@ class Node:
         :param when: The conditional attached to the node
         :param raw_object: The raw ansible object matching this node in the graph. Will be None if there is no match on
         Ansible side
+        :param parent: The parent of this node
         """
         self.name = node_name
+        self.parent = parent
         self.id = node_id
         self.when = when
         self.raw_object = raw_object
@@ -49,6 +50,20 @@ class Node:
         """
         if self.raw_object and self.raw_object.get_ds():
             self.path, self.line, self.column = self.raw_object.get_ds().ansible_pos
+
+    def get_first_parent_matching_type(self, node_type: Type) -> 'Type':
+        """
+        Get the first parent of this node matching the given type
+        :return:
+        """
+        current_parent = self.parent
+
+        while current_parent is not None:
+            if isinstance(current_parent, node_type):
+                return current_parent
+            current_parent = current_parent.parent
+
+        raise ValueError(f"No parent of type {node_type} found for {self}")
 
     def __repr__(self):
         return f"{type(self).__name__}(name='{self.name}',id='{self.id}')"
@@ -74,6 +89,7 @@ class CompositeNode(Node):
         node_id: str,
         when: str = "",
         raw_object=None,
+        parent: 'Node' = None,
         supported_compositions: List[str] = None,
     ):
         """
@@ -84,7 +100,7 @@ class CompositeNode(Node):
         Ansible side
         :param supported_compositions:
         """
-        super().__init__(node_name, node_id, when, raw_object)
+        super().__init__(node_name, node_id, when, raw_object, parent)
         self._supported_compositions = supported_compositions or []
         # The dict will contain the different types of composition.
         self._compositions = defaultdict(list)  # type: Dict[str, List]
@@ -133,17 +149,17 @@ class CompositeNode(Node):
                 elif isinstance(node, CompositeNode):
                     node._get_all_tasks_nodes(task_acc)
 
-    def links_structure(self) -> Dict[str, List[Node]]:
+    def links_structure(self) -> Dict[Node, List[Node]]:
         """
         Return a representation of the composite node where each key of the dictionary is the node id and the
          value is the list of the linked nodes
         :return:
         """
-        links: Dict[str, List[Node]] = defaultdict(list)
+        links: Dict[Node, List[Node]] = defaultdict(list)
         self._get_all_links(links)
         return links
 
-    def _get_all_links(self, links: Dict[str, List[Node]]):
+    def _get_all_links(self, links: Dict[Node, List[Node]]):
         """
         Recursively get the node links
         :return:
@@ -152,7 +168,7 @@ class CompositeNode(Node):
             for node in nodes:
                 if isinstance(node, CompositeNode):
                     node._get_all_links(links)
-                links[self.id].append(node)
+                links[self].append(node)
 
 
 class CompositeTasksNode(CompositeNode):
@@ -160,8 +176,8 @@ class CompositeTasksNode(CompositeNode):
     A special composite node which only support adding "tasks". Useful for block and role
     """
 
-    def __init__(self, node_name: str, node_id: str, when: str = "", raw_object=None):
-        super().__init__(node_name, node_id, when=when, raw_object=raw_object)
+    def __init__(self, node_name: str, node_id: str, when: str = "", raw_object=None, parent: 'Node' = None):
+        super().__init__(node_name, node_id, when=when, raw_object=raw_object, parent=parent)
         self._supported_compositions = ["tasks"]
 
     def add_node(self, target_composition: str, node: Node):
@@ -216,28 +232,31 @@ class PlaybookNode(CompositeNode):
         """
         return self._compositions["plays"]
 
-    def roles_usage(self) -> Dict["RoleNode", List[str]]:
+    def roles_usage(self) -> Dict["RoleNode", Set[Node]]:
         """
         For each role in the graph, return the plays that reference the role
         FIXME: Review this implementation. It may not be the most efficient way, but it's ok for the moment
         :return: A dict with key as role node and value the list of plays
         """
 
-        usages = defaultdict(list)
+        usages = defaultdict(set)
         links = self.links_structure()
 
-        for node_id, linked_nodes in links.items():
+        for node, linked_nodes in links.items():
             for linked_node in linked_nodes:
                 if isinstance(linked_node, RoleNode):
-                    usages[linked_node].append(node_id)
+                    if isinstance(node, PlayNode):
+                        usages[linked_node].add(node)
+                    else:
+                        usages[linked_node].add(node.get_first_parent_matching_type(PlayNode))
 
         # In case a role is used by another role, replace it by the play associated with using role (transitivity)
         for usages_set in usages.values():
-            for node_id in usages_set.copy():
+            for node in usages_set.copy():
                 for r in usages:
-                    if node_id == r.id:
-                        usages_set.remove(node_id)
-                        usages_set.extend(usages[r])
+                    if node.id == r.id:
+                        usages_set.remove(node)
+                        usages_set.update(usages[r])
 
         return usages
 
@@ -257,6 +276,7 @@ class PlayNode(CompositeNode):
         node_id: str = None,
         when: str = "",
         raw_object=None,
+        parent: 'Node' = None,
         hosts: List[str] = None,
     ):
         """
@@ -269,6 +289,7 @@ class PlayNode(CompositeNode):
             node_id or generate_id("play_"),
             when=when,
             raw_object=raw_object,
+            parent=parent,
             supported_compositions=["pre_tasks", "roles", "tasks", "post_tasks"],
         )
         self.hosts = hosts or []
@@ -296,13 +317,14 @@ class BlockNode(CompositeTasksNode):
     """
 
     def __init__(
-        self, node_name: str, node_id: str = None, when: str = "", raw_object=None
+        self, node_name: str, node_id: str = None, when: str = "", raw_object=None, parent: 'Node' = None
     ):
         super().__init__(
             node_name,
             node_id or generate_id("block_"),
             when=when,
             raw_object=raw_object,
+            parent=parent,
         )
 
 
@@ -312,7 +334,7 @@ class TaskNode(Node):
     """
 
     def __init__(
-        self, node_name: str, node_id: str = None, when: str = "", raw_object=None
+        self, node_name: str, node_id: str = None, when: str = "", raw_object=None, parent: 'Node' = None
     ):
         """
 
@@ -321,7 +343,7 @@ class TaskNode(Node):
         :param raw_object:
         """
         super().__init__(
-            node_name, node_id or generate_id("task_"), when=when, raw_object=raw_object
+            node_name, node_id or generate_id("task_"), when=when, raw_object=raw_object, parent=parent
         )
 
 
@@ -336,6 +358,7 @@ class RoleNode(CompositeTasksNode):
         node_id: str = None,
         when: str = "",
         raw_object=None,
+        parent: 'Node' = None,
         include_role: bool = False,
     ):
         """
@@ -346,7 +369,7 @@ class RoleNode(CompositeTasksNode):
         """
         self.include_role = include_role
         super().__init__(
-            node_name, node_id or generate_id("role_"), when=when, raw_object=raw_object
+            node_name, node_id or generate_id("role_"), when=when, raw_object=raw_object, parent=parent
         )
 
     def retrieve_position(self):

--- a/ansibleplaybookgrapher/graphbuilder.py
+++ b/ansibleplaybookgrapher/graphbuilder.py
@@ -12,7 +12,7 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
-from typing import Dict, Optional, Tuple, List
+from typing import Dict, Optional, Tuple, List, Set
 
 from ansible.utils.display import Display
 from graphviz import Digraph
@@ -146,7 +146,7 @@ class GraphvizGraphBuilder:
         open_protocol_handler: str,
         digraph: Digraph,
         play_colors: Dict[PlayNode, Tuple[str, str]],
-        roles_usage: Dict[RoleNode, List[str]] = None,
+        roles_usage: Dict[RoleNode, Set[Node]] = None,
         roles_built: Dict = None,
         open_protocol_custom_formats: Dict[str, str] = None,
     ):
@@ -337,9 +337,13 @@ class GraphvizGraphBuilder:
         if role_to_render is None:
             # Merge the colors for each play where this role is used
             role_plays = self.roles_usage[destination]
-            colors = list(map(self.play_colors.get, role_plays))
             # Graphviz support providing multiple colors separated by :
-            role_color = colors[0][0]
+            if len(role_plays) > 1:
+                # If the role is used in multiple plays, we take black as the default color
+                role_color = "black"
+            else:
+                colors = list(map(self.play_colors.get, role_plays))[0]
+                role_color = colors[0]
 
             self.roles_built[destination.name] = destination
 

--- a/ansibleplaybookgrapher/graphbuilder.py
+++ b/ansibleplaybookgrapher/graphbuilder.py
@@ -22,7 +22,8 @@ from ansibleplaybookgrapher.graph import (
     PlaybookNode,
     RoleNode,
     BlockNode,
-    Node, PlayNode,
+    Node,
+    PlayNode,
 )
 from ansibleplaybookgrapher.utils import get_play_colors, merge_dicts
 
@@ -146,7 +147,7 @@ class GraphvizGraphBuilder:
         open_protocol_handler: str,
         digraph: Digraph,
         play_colors: Dict[PlayNode, Tuple[str, str]],
-        roles_usage: Dict[RoleNode, Set[Node]] = None,
+        roles_usage: Dict[RoleNode, List[Node]] = None,
         roles_built: Dict = None,
         open_protocol_custom_formats: Dict[str, str] = None,
     ):

--- a/ansibleplaybookgrapher/graphbuilder.py
+++ b/ansibleplaybookgrapher/graphbuilder.py
@@ -22,7 +22,7 @@ from ansibleplaybookgrapher.graph import (
     PlaybookNode,
     RoleNode,
     BlockNode,
-    Node,
+    Node, PlayNode,
 )
 from ansibleplaybookgrapher.utils import get_play_colors, merge_dicts
 
@@ -86,7 +86,7 @@ class Grapher:
             for play in playbook_node.plays:
                 # TODO: find a way to create visual distance between the generated colors
                 #   https://stackoverflow.com/questions/9018016/how-to-compare-two-colors-for-similarity-difference
-                self.plays_color[play.id] = get_play_colors(play.id)
+                self.plays_color[play] = get_play_colors(play.id)
 
             # Update the usage of the roles
             self.roles_usage = merge_dicts(
@@ -142,11 +142,11 @@ class GraphvizGraphBuilder:
 
     def __init__(
         self,
-        playbook_node: "PlaybookNode",
+        playbook_node: PlaybookNode,
         open_protocol_handler: str,
         digraph: Digraph,
-        play_colors: Dict[str, Tuple[str, str]],
-        roles_usage: Dict["RoleNode", List[str]] = None,
+        play_colors: Dict[PlayNode, Tuple[str, str]],
+        roles_usage: Dict[RoleNode, List[str]] = None,
         roles_built: Dict = None,
         open_protocol_custom_formats: Dict[str, str] = None,
     ):
@@ -339,7 +339,7 @@ class GraphvizGraphBuilder:
             role_plays = self.roles_usage[destination]
             colors = list(map(self.play_colors.get, role_plays))
             # Graphviz support providing multiple colors separated by :
-            role_color = ":".join([c[0] for c in colors])
+            role_color = colors[0][0]
 
             self.roles_built[destination.name] = destination
 
@@ -361,8 +361,6 @@ class GraphvizGraphBuilder:
                         counter=role_task_counter,
                         color=role_color,
                     )
-        else:
-            print("here")
 
     def build_graphviz_graph(self):
         """
@@ -380,7 +378,7 @@ class GraphvizGraphBuilder:
 
         for play_counter, play in enumerate(self.playbook_node.plays, 1):
             with self.digraph.subgraph(name=play.name) as play_subgraph:
-                color, play_font_color = self.play_colors[play.id]
+                color, play_font_color = self.play_colors[play]
                 play_tooltip = (
                     ",".join(play.hosts) if len(play.hosts) > 0 else play.name
                 )

--- a/ansibleplaybookgrapher/parser.py
+++ b/ansibleplaybookgrapher/parser.py
@@ -118,6 +118,7 @@ class BaseParser(ABC):
                 generate_id(f"{node_type}_"),
                 when=convert_when_to_str(task.when),
                 raw_object=task,
+                parent=parent_node
             ),
         )
 
@@ -229,6 +230,7 @@ class PlaybookParser(BaseParser):
                     clean_name(role.get_name()),
                     node_id="role_" + hash_value(role.get_name()),
                     raw_object=role,
+                    parent=play_node,
                 )
                 # edge from play to role
                 play_node.add_node("roles", role_node)
@@ -297,7 +299,7 @@ class PlaybookParser(BaseParser):
         if not block._implicit and block._role is None:
             # Here we have an explicit block. Ansible internally converts all normal tasks to Block
             block_node = BlockNode(
-                str(block.name), when=convert_when_to_str(block.when), raw_object=block
+                str(block.name), when=convert_when_to_str(block.when), raw_object=block, parent=parent_nodes[-1]
             )
             parent_nodes[-1].add_node(f"{node_type}s", block_node)
             parent_nodes.append(block_node)
@@ -341,6 +343,7 @@ class PlaybookParser(BaseParser):
                         node_id="role_" + hash_value(task_or_block._role_name),
                         when=convert_when_to_str(task_or_block.when),
                         raw_object=task_or_block,
+                        parent=parent_nodes[-1],
                         include_role=True,
                     )
                     parent_nodes[-1].add_node(

--- a/ansibleplaybookgrapher/parser.py
+++ b/ansibleplaybookgrapher/parser.py
@@ -118,7 +118,7 @@ class BaseParser(ABC):
                 generate_id(f"{node_type}_"),
                 when=convert_when_to_str(task.when),
                 raw_object=task,
-                parent=parent_node
+                parent=parent_node,
             ),
         )
 
@@ -299,7 +299,10 @@ class PlaybookParser(BaseParser):
         if not block._implicit and block._role is None:
             # Here we have an explicit block. Ansible internally converts all normal tasks to Block
             block_node = BlockNode(
-                str(block.name), when=convert_when_to_str(block.when), raw_object=block, parent=parent_nodes[-1]
+                str(block.name),
+                when=convert_when_to_str(block.when),
+                raw_object=block,
+                parent=parent_nodes[-1],
             )
             parent_nodes[-1].add_node(f"{node_type}s", block_node)
             parent_nodes.append(block_node)

--- a/ansibleplaybookgrapher/postprocessor.py
+++ b/ansibleplaybookgrapher/postprocessor.py
@@ -137,10 +137,10 @@ class GraphVizPostProcessor:
         display.vv(f"Inserting links structure for the playbook '{playbook_node.name}'")
         links_structure = playbook_node.links_structure()
 
-        for node_id, node_links in links_structure.items():
+        for node, node_links in links_structure.items():
             # Find the group g with the specified id
             xpath_result = self.root.xpath(
-                f"ns:g/*[@id='{node_id}']", namespaces={"ns": SVG_NAMESPACE}
+                f"ns:g/*[@id='{node.id}']", namespaces={"ns": SVG_NAMESPACE}
             )
             if xpath_result:
                 element = xpath_result[0]
@@ -151,7 +151,7 @@ class GraphVizPostProcessor:
                             "link",
                             attrib={
                                 "target": link.id,
-                                "edge": f"edge_{counter}_{node_id}_{link.id}",
+                                "edge": f"edge_{counter}_{node.id}_{link.id}",
                             },
                         )
                     )

--- a/ansibleplaybookgrapher/utils.py
+++ b/ansibleplaybookgrapher/utils.py
@@ -18,7 +18,7 @@ import uuid
 from collections import defaultdict
 from itertools import chain
 from operator import methodcaller
-from typing import Tuple, List, Dict, Any
+from typing import Tuple, List, Dict, Any, Set
 
 from ansible.errors import AnsibleError
 from ansible.module_utils._text import to_text
@@ -111,18 +111,18 @@ def has_role_parent(task_block: Task) -> bool:
     return False
 
 
-def merge_dicts(dict_1: Dict[Any, List], dict_2: Dict[Any, List]) -> Dict[Any, List]:
+def merge_dicts(dict_1: Dict[Any, Set], dict_2: Dict[Any, Set]) -> Dict[Any, Set]:
     """
     Merge two dicts by grouping keys and appending values in list
     :param dict_1:
     :param dict_2:
     :return:
     """
-    final = defaultdict(list)
+    final = defaultdict(set)
     # iterate dict items
     all_dict_items = map(methodcaller("items"), [dict_1, dict_2])
     for k, v in chain.from_iterable(all_dict_items):
-        final[k].extend(v)
+        final[k].update(v)
 
     return final
 

--- a/ansibleplaybookgrapher/utils.py
+++ b/ansibleplaybookgrapher/utils.py
@@ -111,18 +111,18 @@ def has_role_parent(task_block: Task) -> bool:
     return False
 
 
-def merge_dicts(dict_1: Dict[Any, Set], dict_2: Dict[Any, Set]) -> Dict[Any, Set]:
+def merge_dicts(dict_1: Dict[Any, List], dict_2: Dict[Any, List]) -> Dict[Any, List]:
     """
     Merge two dicts by grouping keys and appending values in list
     :param dict_1:
     :param dict_2:
     :return:
     """
-    final = defaultdict(set)
+    final = defaultdict(list)
     # iterate dict items
     all_dict_items = map(methodcaller("items"), [dict_1, dict_2])
     for k, v in chain.from_iterable(all_dict_items):
-        final[k].update(v)
+        final[k].extend(v)
 
     return final
 

--- a/tests/fixtures/include_role.yml
+++ b/tests/fixtures/include_role.yml
@@ -6,12 +6,13 @@
     - fake_role
     - display_some_facts
   tasks:
-    - name: (0) First Include role (with loop)
-      include_role:
-        name: '{{ item }}'
-      loop:
-        - fake_role
-        - display_some_facts
+    - block:
+        - name: (0) First Include role (with loop)
+          include_role:
+            name: '{{ item }}'
+          loop:
+            - fake_role
+            - display_some_facts
 
     - name: (1) Debug
       debug:

--- a/tests/fixtures/roles/nested_include_role/tasks/main.yml
+++ b/tests/fixtures/roles/nested_include_role/tasks/main.yml
@@ -10,10 +10,9 @@
     state: started
 
 - name: Include role block
-  block:
-    - when: x is not defined
-      include_role:
-        name: display_some_facts
+  when: x is not defined
+  include_role:
+    name: display_some_facts
 
 - name: Include role
   include_role:

--- a/tests/fixtures/roles/nested_include_role/tasks/main.yml
+++ b/tests/fixtures/roles/nested_include_role/tasks/main.yml
@@ -9,10 +9,11 @@
     name: postgresql
     state: started
 
-- name: Include role
-  when: x is not defined
-  include_role:
-    name: display_some_facts
+- name: Include role block
+  block:
+    - when: x is not defined
+      include_role:
+        name: display_some_facts
 
 - name: Include role
   include_role:

--- a/tests/test_graph.py
+++ b/tests/test_graph.py
@@ -26,17 +26,17 @@ def test_links_structure():
     play.add_node("tasks", task_3)
 
     all_links = play.links_structure()
-    assert len(all_links) == 2, "The links should contains only 6 elements"
+    assert len(all_links) == 2, "The links should contains only 2 elements"
 
-    assert len(all_links[play.id]) == 2, "The play should be linked to 2 nodes"
+    assert len(all_links[play]) == 2, "The play should be linked to 2 nodes"
     for e in [role, task_3]:
         assert (
-            e in all_links[play.id]
+            e in all_links[play]
         ), f"The play should be linked to the task {task_1}"
 
-    assert len(all_links[role.id]) == 2, "The role should be linked to two nodes"
+    assert len(all_links[role]) == 2, "The role should be linked to two nodes"
     for e in [task_1, task_2]:
-        assert e in all_links[role.id], f"The role should be linked to the edge {e}"
+        assert e in all_links[role], f"The role should be linked to the edge {e}"
 
 
 def test_get_all_tasks_nodes():

--- a/tests/test_graph.py
+++ b/tests/test_graph.py
@@ -30,9 +30,7 @@ def test_links_structure():
 
     assert len(all_links[play]) == 2, "The play should be linked to 2 nodes"
     for e in [role, task_3]:
-        assert (
-            e in all_links[play]
-        ), f"The play should be linked to the task {task_1}"
+        assert e in all_links[play], f"The play should be linked to the task {task_1}"
 
     assert len(all_links[role]) == 2, "The role should be linked to two nodes"
     for e in [task_1, task_2]:

--- a/tests/test_parser.py
+++ b/tests/test_parser.py
@@ -228,13 +228,18 @@ def test_roles_usage(grapher_cli: PlaybookGrapherCLI):
         assert all(
             map(lambda node: node.id.startswith("play_"), plays)
         ), "All nodes IDs should be play"
-        nb_plays = len(plays)
+
+        nb_plays_for_the_role = len(plays)
 
         if role.name == "fake_role":
-            assert nb_plays == 3, "The role fake_role is used 3 times in the plays"
+            assert (
+                nb_plays_for_the_role == 3
+            ), "The role fake_role is used 3 times in the plays"
         elif role.name == "display_some_facts":
             assert (
-                nb_plays == 4
+                nb_plays_for_the_role == 4
             ), "The role display_some_facts is used 4 times in the plays"
         elif role.name == "nested_included_role":
-            assert nb_plays == 1, "The role nested_included_role is used in 1 play"
+            assert (
+                nb_plays_for_the_role == 1
+            ), "The role nested_included_role is used in 1 play"

--- a/tests/test_parser.py
+++ b/tests/test_parser.py
@@ -108,12 +108,14 @@ def test_include_role_parsing(grapher_cli: PlaybookGrapherCLI, capsys):
         in capsys.readouterr().err
     ), "A warning should be displayed regarding loop being not supported"
 
-    # first include_role
-    include_role_1 = tasks[0]
+    # first include_role using a block
+    block_include_role = tasks[0]
+    assert isinstance(block_include_role, BlockNode)
+    include_role_1 = block_include_role.tasks[0]
     assert isinstance(include_role_1, RoleNode)
     assert include_role_1.include_role
     assert include_role_1.path == os.path.join(FIXTURES_PATH, "include_role.yml")
-    assert include_role_1.line == 9, "The first include role should be at line 9"
+    assert include_role_1.line == 10, "The first include role should be at line 9"
     assert (
         len(include_role_1.tasks) == 0
     ), "We don't support adding tasks from include_role with loop"

--- a/tests/test_parser.py
+++ b/tests/test_parser.py
@@ -226,7 +226,7 @@ def test_roles_usage(grapher_cli: PlaybookGrapherCLI):
 
     for role, plays in roles_usage.items():
         assert all(
-            map(lambda node_id: node_id.startswith("play_"), plays)
+            map(lambda node: node.id.startswith("play_"), plays)
         ), "All nodes IDs should be play"
         nb_plays = len(plays)
 

--- a/tests/test_playbook_grapher.py
+++ b/tests/test_playbook_grapher.py
@@ -241,6 +241,7 @@ def test_include_role(request, include_role_tasks_option, expected_tasks_number)
         svg_path=svg_path,
         playbook_paths=playbook_paths,
         plays_number=1,
+        blocks_number=1,
         tasks_number=expected_tasks_number,
         roles_number=3,
     )

